src/algebra/quandle.lean

```diff
@@ -0,0 +1,561 @@
+* Alexander quandle with `a ◃ b = t * b + (1 - t) * b`, with `a` and `b` elements of a module over Z[t,t⁻¹].
+* If G is a group, H a subgroup, and z in H, then there is a quandle `(G/H;z)` defined by
+  `yH ◃ Hx = yzy⁻¹xH`.  Every homogeneous quandle (i.e., a quandle Q whose automorphism group acts
+  transitively on Q as a set) is isomorphic to such a quandle.
+  There is a generalization to this arbitrary quandles in [Joyce's paper (Theorem 7.2)][Joyce1982].
+
+## Tags
+
+rack, quandle
+-/
+
+universes u v
+
+/--
+A *rack* is an automorphic set (a set with an action on itself) that is self-distributive.
+The `x ◃ y` and `x ◃⁻¹ y` notation is right associative.
+
+There is a weaker notion of a *shelf*, which would be from having `op' : α → α → α`.
+-/
+class rack (α : Type*) :=
+(op' : α → (α ≃ α))
+(self_distrib' : ∀ {x y z : α}, op' x (op' y z) = op' (op' x y) (op' x z))
+
+/--
+The rack operation.  Has notation `◃`.
+-/
+abbreviation rack.op {R : Type*} [rack R] (x y : R) : R := rack.op' x y
+/--
+The inverse of the rack operation. Has notation `◃⁻¹`.
+-/
+abbreviation rack.inv_op {R : Type*} [rack R] (x y : R) : R := (rack.op' x).symm y
+
+local infixr ` ◃ ` : 65 := rack.op
+local infixr ` ◃⁻¹ ` : 65 := rack.inv_op
+
+namespace rack
+variables {R : Type*} [rack R]
+
+@[simp] lemma normalize_op (x y : R) : op' x y = x ◃ y := rfl
+@[simp] lemma normalize_inv_op (x y : R) : (op' x).symm y = x ◃⁻¹ y := rfl
+@[simp] lemma normalize_inv_op' (x y : R) : (op' x)⁻¹ y = x ◃⁻¹ y := rfl
+
+@[simp] lemma left_inv (x y : R) : x ◃⁻¹ x ◃ y = y := (op' x).symm_apply_eq.mpr rfl
+@[simp] lemma right_inv (x y : R) : x ◃ x ◃⁻¹ y = y := (op' x).eq_symm_apply.mp rfl
+
+lemma self_distrib {x y z : R} : x ◃ y ◃ z = (x ◃ y) ◃ (x ◃ z) :=
+rack.self_distrib'
+
+lemma left_cancel (x : R) {y y' : R} : x ◃ y = x ◃ y' ↔ y = y' :=
+by { split, apply (op' x).injective, rintro rfl, refl }
+
+lemma left_cancel_inv (x : R) {y y' : R} : x ◃⁻¹ y = x ◃⁻¹ y' ↔ y = y' :=
+by { split, apply (op' x).symm.injective, rintro rfl, refl }
+
+lemma self_distrib_inv {x y z : R} : x ◃⁻¹ y ◃⁻¹ z = (x ◃⁻¹ y) ◃⁻¹ (x ◃⁻¹ z) :=
+begin
+  rw [←left_cancel (x ◃⁻¹ y), right_inv, ←left_cancel x, right_inv, self_distrib],
+  repeat {rw right_inv },
+end
+
+/--
+The *adjoint action* of a rack on itself is `op'`, and the adjoint
+action of `x ◃ y` is the conjugate of the action of `y` by the action
+of `x`. It is another way to understand the self-distributivity axiom.
+
+This is used in the natural rack homomorphism `to_conj` from `R` to
+`conj (R ≃ R)` defined by `op'`.
+-/
+lemma ad_conj {R : Type*} [rack R] (x y : R) :
+  op' (x ◃ y) = op' x * op' y * (op' x)⁻¹ :=
+begin
+  apply @mul_right_cancel _ _ _ (op' x), ext z,
+  simp only [inv_mul_cancel_right],
+  apply self_distrib.symm,
+end
+
+/--
+The opposite rack, swapping the roles of `◃` and `◃⁻¹`.
+-/
+@[nolint has_inhabited_instance]
+def opp (R : Type*) := R
```

Comment:
Does it make sense to just use the existing `op`, which we use for opposite monoids, opposite rings, opposite categories, etc?

Reply:
Yes, `op` seems best. Thanks